### DISCUSSION TITLE
Resource editor - use consistent uris

### DISCUSSION THREAD

**Discussion by armin11** [mrmap-community/mrmap#88] [Ideas]
### Environment
* Python version: 3.7.9
* MrMap version: pre1.0

### Proposed Functionality
Use a strict uri concept for the resource editor when managing services

### Use Case
e.g.
{mrmap}resource/wms/

Try to activate one service by pushing button:

https://mrmap.geospatial-interoperability-solutions.eu/resource/20b2cef2-fd69-493d-a5d1-c741b99c4030/activate

The form to activate the resource is shown - set checkbox and push save will redirect to resource service:

https://mrmap.geospatial-interoperability-solutions.eu/resource/20b2cef2-fd69-493d-a5d1-c741b99c4030

It will be better to get back to the https://mrmap.geospatial-interoperability-solutions.eu/resource/wms/ list.

Maybe this can be done by setting some header/session information from where the invocation of the edit came from. Or it will be done directly without go thru https://mrmap.geospatial-interoperability-solutions.eu/resource/20b2cef2-fd69-493d-a5d1-c741b99c4030/activate - but only invoke s.th. like https://mrmap.geospatial-interoperability-solutions.eu/resource/20b2cef2-fd69-493d-a5d1-c741b99c4030/activate/true wit a redirect to the calling page on success.

Similar:
Link on update: https://mrmap.geospatial-interoperability-solutions.eu/resource/run-update/20b2cef2-fd69-493d-a5d1-c741b99c4030 - this is not in line with https://mrmap.geospatial-interoperability-solutions.eu/resource/20b2cef2-fd69-493d-a5d1-c741b99c4030/{operation} as it is done for activate, remove, edit-metadata, access-edit

And:
Link on run monitor: https://mrmap.geospatial-interoperability-solutions.eu/monitoring/runs/new?metadatas=35fd285f-bac9-4e88-8479-b1b82ea68765 links to monitor for metadata - maybe also a {mrmap}/resource/{uuid}/monitor/start uri maybe better?

Proposed list of operations (maybe "resource-" is not needed):
* resource-{activate}
* resource-{edit-metadata}
* resource-{update}
* resource-{monitor}
* resource-{edit-access}
* resource-{delete}

Compare ckan api (it has an similar information model): https://docs.ckan.org/en/2.9/api/ - to discuss ;-)

### Database Changes
none
### External Dependencies
none

**Comment by jokiefer** [comment]
Moved to discussion, cause not all of your proposals are clear enough to implement them straightforward.

**Comment by jokiefer** [comment]
> Try to activate one service by pushing button:
https://mrmap.geospatial-interoperability-solutions.eu/resource/20b2cef2-fd69-493d-a5d1-c741b99c4030/activate
The form to activate the resource is shown 

this behavior only occurs in the list wms/wfs/csw views. The form shall not viewed. Take a look on the resource detail view to see the right behavior. 
--> One housekeeping issue to change the activate/deactivate button in rendered tables from link to form with instant committing like in the resource detail view.

>  set checkbox and push save will redirect to resource service:
https://mrmap.geospatial-interoperability-solutions.eu/resource/20b2cef2-fd69-493d-a5d1-c741b99c4030
It will be better to get back to the https://mrmap.geospatial-interoperability-solutions.eu/resource/wms/ list.

This behavior is the default behavior of class based views ([UpdateView](https://github.com/django/django/blob/main/django/views/generic/edit.py#L110))
--> new housekeeping issue: overwrite [get_success_url()](https://github.com/mrmap-community/mrmap/blob/master/mrmap/service/views.py#L204) of `ResourceActivateDeactivateView` class. We need to discuss how we tell the `ResourceActivateDeactivateView` view where we come from.

>Similar:
Link on update: https://mrmap.geospatial-interoperability-solutions.eu/resource/run-update/20b2cef2-fd69-493d-a5d1-c741b99c4030 - this is not in line with https://mrmap.geospatial-interoperability-solutions.eu/resource/20b2cef2-fd69-493d-a5d1-c741b99c4030/{operation} as it is done for activate, remove, edit-metadata, access-edit

That's right. The update views are not refactored as class based views and there url are also not touched. We need to refactor them as class based views and think about the urls. See issue [#14](https://github.com/mrmap-community/mrmap/issues/14#issuecomment-815179703) (i added a comment)

>And:
Link on run monitor: https://mrmap.geospatial-interoperability-solutions.eu/monitoring/runs/new?metadatas=35fd285f-bac9-4e88-8479-b1b82ea68765 links to monitor for metadata - maybe also a {mrmap}/resource/{uuid}/monitor/start uri maybe better?

That not a common way to create MonitoringRun objects with a generic django class based view. [See](https://github.com/mrmap-community/mrmap/blob/master/mrmap/monitoring/views.py#L26) for better understanding how the views is configured. It uses the [CreateView](https://github.com/django/django/blob/main/django/views/generic/edit.py#L175) from django to create the MonitoringRun object. If we set up the uri  as you recommended, we have to build up a non generic custom view (maybe function based instead of class based) which receives the uuid of the resource and creates a MonitoringRun object custom coded by us. Maybe we could jump over the form step like the activate/deactivate behavior to start the task directly.

>Proposed list of operations (maybe "resource-" is not needed):
resource-{activate}
resource-{edit-metadata}
resource-{update}
resource-{monitor}
resource-{edit-access}
resource-{delete}

principal good. But again take a look on monitor.

>Compare ckan api (it has an similar information model): https://docs.ckan.org/en/2.9/api/ - to discuss ;-)

Do not think to complex. The frontend of mr map is not an api, where we really need to define well known uris to perform actions, cause all actions you can perform are linked in html elements like buttons and so on. In my opinion we should participial construct uris like `{app}/{kind-of-objects}/{pk}/{action}.
